docs/prefetch.md

```diff
@@ -0,0 +1,112 @@
+They have different sizes, but no greater than 100. 
+
+#### PrefetchContextProvider
+
+It wraps all components, on start of the runtime it sets up the queue but leaves if turned off. After receiving the load signal from the browser, it waits a few extra seconds, then turns on the queue and the requests starts. We had to do it because otherwise, the donwload of prefetched components was affecting lighthouse scores.
+It also waits for navigation events, when they happen, we clear the queue, and pause it for one second to allow the browser to have some time to request assets and have more free network and connections.
+
+### Prefetch Hook
+
+Its located at react/hooks/prefetch.ts
+
+Before trying to make requests the hook does some logic, including:
+- Call navigation route modifiers to parse the given path and get a more uniform result. For example: /Shoes and /shoes are normalized to /shoes, increasing cache hits.
+- Check if for that given path, we already have all data required in our LRU caches (page and path data). If we need any data, move on to add request to queue.
+
+#### Requests logic
+
+At this moment we were already started by the promise queue.
+We first try to get the path data from the correct LRU path cache. If not available, we will make a request to rewriter => render-server.
+This request is done in the structure: `/PATH?__pickRuntime=page,queryData,contentResponse,route`.
+For that path we get: the page id for that path (store.product), the query data (the result of the product query, for example), route data (the variables, canonical, built by render-server and pages) and the content response that has data custom data for that path, like content and components.
+We then save this data on the appropriate LRU path cache on prefetch state.
+
+After the request is done we now have all data that varies by path.
+We can then, if not in cache already, try to fetch data for that page.
+To fetch the page data, we do a query on pages-graphql.
+
+```
+query RouteData($routeId: String, $declarer: String, $query: String, $device: UserDevice, $renderMajor: Int) {
+  prefetchBlocks(input:{ routeId:$routeId, renderMajor:$renderMajor, declarer: $declarer, query: $query, device: $device}) {
+    extensionsJSON
+    componentsJSON
+    messages {
+      key
+      message
+    }
+  }
+}
+```
+
+This query is done with the `no-cache` fetch policy. This means we will not store the result in Apollo-cache (we are already storing in our own cache).
+This is necessary otherwise we would save a lot of data and consume too mch memory from our user device.
+After this query is done, it is saved in our own LRU cache (routesCache). We store the messages, the extensions for that page and the components for it.
+
+Finally, we can fetch the assets for those components on background (this consumes cpu since we are donwloading Javascript that is processed as it is being downloaded).
+
+This is the end of the "background async" part of prefetch.
```

Comment:
```suggestion
Finally, we can fetch the assets for those components on the background (this consumes CPU since we are downloading Javascript that is processed as it is being downloaded).

This is the end of the "background async" part of prefetch.
```